File: typescript/src/OnCall/OnCallManager.ts
/**
 * Zirve OnCall Manager — Grafana OnCall
 *
 * Interacts with Grafana OnCall API to create alerts and manage incidents.
 */
export class OnCallManager {
  private config: Record<string, any>;
  private url: string;
  private token: string;

  constructor(config: Record<string, any>) {
    this.config = config;
    this.url = (config.url || 'http://grafana-oncall-engine.zirve-infra.svc.cluster.local:8080').replace(/\/$/, '') + '/api/v1';
    this.token = config.token || '';
  }

  private async request(method: string, path: string, body?: any): Promise<any> {
    const options: RequestInit = {
      method,
      headers: {
        'Authorization': `Token ${this.token}`,
        'Content-Type': 'application/json'
      },
      body: body ? JSON.stringify(body) : undefined
    };

    const response = await fetch(`${this.url}/${path.replace(/^\/+/, '')}`, options);
    
    if (!response.ok) {
      if (response.status === 404) return null;
      throw new Error(`Grafana OnCall API Error [${response.status}]`);
    }

    if (response.status === 204 || response.headers.get('content-length') === '0') return true;
    return await response.json();
  }

  public async createAlert(integrationUrl: string, title: string, message: string, state: 'alerting' | 'ok' = 'alerting'): Promise<boolean> {
    // Note: Grafana OnCall uses an integration URL for inbound webhooks
    try {
      const response = await fetch(integrationUrl, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          title,
          message,
          state
        })
      });
      return response.ok;
    } catch {
      return false;
    }
  }

  public async listSchedules(): Promise<any[]> {
    const result = await this.request('GET', 'schedules');
    return result?.results || [];
  }

  public async listIncidents(state: 'triggered' | 'acknowledged' | 'resolved' = 'triggered'): Promise<any[]> {
    const result = await this.request('GET', `alert_groups?state=${state}`);
    return result?.results || [];
  }

  public async health(): Promise<boolean> {
    try {
      const response = await fetch(`${this.url}/health`);
      return response.ok;
    } catch {
      return false;
    }
  }
}

File: dotnet/Zirve.Sdk/OnCall/OnCallManager.cs
using System;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;
using Zirve.Sdk.Config;

namespace Zirve.Sdk.OnCall;

public class OnCallManager
{
    private readonly HttpClient _httpClient;
    private readonly string _url;
    private readonly string _token;

    public OnCallManager(HttpClient httpClient, ConfigManager configManager)
    {
        _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
        
        var cfg = configManager.Module("oncall");
        _url = (cfg.GetValueOrDefault("url", "http://grafana-oncall-engine.zirve-infra.svc.cluster.local:8080") ?? "").TrimEnd('/') + "/api/v1";
        _token = cfg.GetValueOrDefault("token", "") ?? "";
    }

    private async Task<JsonElement> RequestAsync(HttpMethod method, string path, object? body = null)
    {
        var request = new HttpRequestMessage(method, $"{_url}/{path.TrimStart('/')}");
        request.Headers.Add("Authorization", $"Token {_token}");
        
        if (body != null)
        {
            var json = body is string s ? s : JsonSerializer.Serialize(body);
            request.Content = new StringContent(json, System.Text.Encoding.UTF8, "application/json");
        }

        var response = await _httpClient.SendAsync(request);
        
        if (response.StatusCode == System.Net.HttpStatusCode.NotFound)
            return default;

        response.EnsureSuccessStatusCode();

        if (response.StatusCode == System.Net.HttpStatusCode.NoContent || response.Content.Headers.ContentLength == 0)
            return default;

        using var doc = await JsonDocument.ParseAsync(await response.Content.ReadAsStreamAsync());
        return doc.RootElement.Clone();
    }

    public async Task<bool> CreateAlertAsync(string integrationUrl, string title, string message, string state = "alerting")
    {
        try
        {
            var request = new HttpRequestMessage(HttpMethod.Post, integrationUrl)
            {
                Content = new StringContent(JsonSerializer.Serialize(new
                {
                    title,
                    message,
                    state
                }), System.Text.Encoding.UTF8, "application/json")
            };
            
            var response = await _httpClient.SendAsync(request);
            return response.IsSuccessStatusCode;
        }
        catch
        {
            return false;
        }
    }

    public async Task<JsonElement> ListIncidentsAsync(string state = "triggered")
    {
        var res = await RequestAsync(HttpMethod.Get, $"alert_groups?state={Uri.EscapeDataString(state)}");
        if (res.ValueKind == JsonValueKind.Object && res.TryGetProperty("results", out var results))
        {
            return results;
        }
        return res;
    }

    public async Task<bool> HealthAsync()
    {
        try
        {
            var response = await _httpClient.GetAsync($"{_url}/health");
            return response.IsSuccessStatusCode;
        }
        catch
        {
            return false;
        }
    }
}
